### PR TITLE
HeaderHelper: add getParameters(), fix getValuesAndParameters()

### PR DESCRIPTION
**Pull request by kamarton**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ✔️
| Breaks BC?    | ❌
| Tests pass?   | ✔️
| Fixed PR comment  | https://github.com/yiisoft/yii-web/pull/157#discussion_r351734462
| Related PR | https://github.com/yiisoft/yii-web/pull/157

### CODE REVIEW

**Comment by samdark** [comment]
Merged. Thanks!

File: src/Helper/HeaderHelper.php [hunk 1]
```diff
@@ -17,16 +17,38 @@ public static function getValueAndParameters(string $headerValue): array
...
+     *
+     * @link https://tools.ietf.org/html/rfc7230#section-3.2.6
+     */
+    public static function getParameters(string $headerValue): array
+    {
+        $headerValue = trim($headerValue);
+        if ($headerValue === '') {
+            return [];
+        }
+        $output = new \stdClass();
```
Comment (xepozz) [inline src/Helper/HeaderHelper.php]: I think it should be simply ```$output = [];```

File: src/Helper/HeaderHelper.php [hunk 2]
```diff
@@ -17,16 +17,38 @@ public static function getValueAndParameters(string $headerValue): array
...
+        $output->data = [];
+        do {
+            $headerValue = preg_replace_callback(
+                '/^\s*(?<parameter>\w+)\s*=\s*(?:"(?<valueQuoted>[^"]+)"|(?<value>[!#$%&\'*+.^`|~\w-]+))\s*(?:;|$)/',
```
Comment (roxblnfk) [inline src/Helper/HeaderHelper.php]: - Some parameter names can contain `*` (e.g. `filename*`)
- Quoted string can contain quoted-pair e.g. `\"`,  `\\` [[Basic Rules]](https://www.rfc-editor.org/rfc/rfc2616.html#section-2.2)
- Some headers can contain a listed values in parameters [[Accept]](https://www.rfc-editor.org/rfc/rfc2616.html#section-14.1), [[Accept-Encoding]](https://www.rfc-editor.org/rfc/rfc2616.html#section-14.3):
`Accept: audio/*; q=0.2, audio/basic`
`Accept-Encoding: gzip;q=1.0, identity; q=0.5, *;q=0`

I think it is impossible to solve this problem with one regular expression.
Moreover, some headlines require a personal touch.

File: src/Helper/HeaderHelper.php (hunk 1 above)
Comment (samdark) [reply src/Helper/HeaderHelper.php]: https://github.com/yiisoft/yii-web/commit/8881b0421485031976912f5c4cc8321b8dde0468

File: src/Helper/HeaderHelper.php (hunk 2 above)
Comment (samdark) [reply src/Helper/HeaderHelper.php]: @roxblnfk is it possible to pull-request a test for it? That would be much easier to fix then.

Comment (kamarton) [reply src/Helper/HeaderHelper.php]: > Some headers can contain a listed values in parameters [Accept], [Accept-Encoding]:

You mean that? https://github.com/yiisoft/yii-web/blob/8881b0421485031976912f5c4cc8321b8dde0468/src/Helper/HeaderHelper.php#L14

> Some parameter names can contain * (e.g. filename*)

in https://tools.ietf.org/html/rfc7230#section-3.2.6:
> Delimiters are chosen from the set of US-ASCII visual characters not allowed in a token (DQUOTE and "(),/:;<=>?@[\]{}").

`*` not allowed, but it's true that the token contains more characters.

> Quoted string can contain quoted-pair e.g. \", \\ [Basic Rules]
https://tools.ietf.org/html/rfc7230#section-3.2.6

agree

Comment (roxblnfk) [reply src/Helper/HeaderHelper.php]: > You mean that?

Yes sure :-| This is more about parsing a list of header values

Comment (kamarton) [reply src/Helper/HeaderHelper.php]: @roxblnfk this is a function that has been used a lot, would you to improve further with new tests in a new PR?

Comment (roxblnfk) [reply src/Helper/HeaderHelper.php]: @kamarton i have other tasks now. I can deal with this problem, but if only after some time

Comment (samdark) [reply src/Helper/HeaderHelper.php]: https://github.com/yiisoft/yii-web/issues/184

**Pull request merged**